packages/components/src/custom-select-control/test/index.js

```diff
@@ -87,4 +88,127 @@ describe( 'CustomSelectControl', () => {
+
+		const newValue = screen.getByRole( 'option', {
+			name: /poppy/i,
+		} );
+
+		await user.click( newValue );
+
+		expect( currentValue ).toHaveTextContent( /poppy/i );
+		expect( mock ).toHaveBeenCalled();
+	} );
+
+	it( 'Should be able to change selection using keyboard', async () => {
+		const user = userEvent.setup();
+
+		render(
+			<CustomSelectControl options={ options } __nextUnconstrainedWidth />
+		);
+
+		const currentValue = screen.getByRole( 'button' );
+		expect( currentValue ).toHaveTextContent( /violets/i );
+
+		await user.tab();
+		expect( currentValue ).toHaveFocus();
+
+		await user.keyboard( '{enter}' );
+		await user.keyboard( '{arrowdown}' );
+		await user.keyboard( '{enter}' );
+
+		expect( currentValue ).toHaveTextContent( /crimson clover/i );
+	} );
+
+	it( 'Should keep current selection if dropdown is closed without changing selection', async () => {
+		const user = userEvent.setup();
+
+		render(
+			<CustomSelectControl options={ options } __nextUnconstrainedWidth />
+		);
+
+		const currentValue = screen.getByRole( 'button' );
+		expect( currentValue ).toHaveTextContent( /violets/i );
+
+		await user.tab();
+		await user.keyboard( '{enter}' );
+		expect( screen.getByRole( 'listbox' ) ).toBeVisible();
+		expect( screen.getByRole( 'listbox' ) ).toHaveAttribute(
+			'aria-hidden',
+			'false'
+		);
+		await user.keyboard( '{escape}' );
+		expect( screen.queryByRole( 'listbox' ) ).not.toBeInTheDocument();
+
+		expect( currentValue ).toHaveTextContent( /violets/i );
+	} );
+
+	it( 'Should be able to type characters to select matching options', async () => {
+		const user = userEvent.setup();
+
+		render(
+			<CustomSelectControl options={ options } __nextUnconstrainedWidth />
+		);
+
+		const currentValue = screen.getByRole( 'button' );
+		expect( currentValue ).toHaveTextContent( /violets/i );
+
+		await user.tab();
+		await user.keyboard( '{enter}' );
+		await user.keyboard( '{p}' );
+		await user.keyboard( '{enter}' );
+
+		expect( currentValue ).toHaveTextContent( /poppy/i );
+	} );
+
+	it( 'Can change selection with a focused input and closed dropdown if typed characters match an option', async () => {
+		const user = userEvent.setup();
+
+		render(
+			<CustomSelectControl options={ options } __nextUnconstrainedWidth />
+		);
+
+		const currentValue = screen.getByRole( 'button' );
+		expect( currentValue ).toHaveTextContent( /violets/i );
+
+		await user.tab();
+		await user.keyboard( '{c}' );
+		await user.keyboard( '{enter}' );
+
+		expect( currentValue ).toHaveTextContent( /crimson clover/i );
+	} );
+
+	it( 'Current selection has attribute: aria-selected="true"', async () => {
+		const user = userEvent.setup();
+
+		render(
+			<CustomSelectControl options={ options } __nextUnconstrainedWidth />
+		);
+
+		await user.tab();
+		await user.keyboard( '{enter}' );
+
+		expect(
+			screen.getByRole( 'option', {
+				name: /violets/i,
+			} )
+		).toHaveAttribute( 'aria-selected', 'true' );
+	} );
 } );
```

Comment:
Maybe it could be nice to see some tests or assertions for:

* Testing the selected value in addition to the selected item text content
* Deselecting, and/or selecting an item without a value.
* Selecting out of multiple options that match a query
* Selecting an option after another option has been selected previously
* Multiple selection?
* Additional item styles
* All additional props that weren't tested/covered here, including the event handler ones

Reply:
Thank you for the suggestions! A couple comments/questions in regards to them:

> Multiple selection?

Not possible currently, but its in the experimental version 😄 so hopefully soon

> Additional item styles

Are you referring to adding styles and classnames to options? Would that just be to ensure that it isn't applied to all items but specific ones? Just wondering how to test it without it being solely about implementation details. 😕

Reply:
Yes! I don't think it's going to be about testing implementation details - it's the opposite, we would be testing the API, and how we can apply specific styles to separate items.

Reply:
Thanks! I've added a couple tests for this in: https://github.com/WordPress/gutenberg/pull/56575/commits/948a131b5ef8e312d4a1436f05fe86daca35c215

Hopefully, I've captured what you suggested, otherwise I'd be curious for your thoughts/feedback.

Reply:
Thanks! 

As mentioned above, it might make sense to add additional assertions for testing the selected value in addition to the selected option text content. Also, what happens if we select an item with an empty value (similar to having an empty string as `value` for an `option` element in HTML)? 

I think you've covered the rest well enough already 👍

Reply:
>As mentioned above, it might make sense to add additional assertions for testing the selected value in addition to the selected option text content. 

Are you referring to adding commented-out assertions that will fail, as [suggested earlier](https://github.com/WordPress/gutenberg/pull/56575#discussion_r1411951559)? Would that help clarify why those assertions aren't there in the first place? 

I'm curious as I was leaning toward the other suggestion - the reason being I think it's likely other things will come up in the new component that we may not have anticipated testing for here. As we bolster the tests in the new component, we may have additional tests/assertions to bring back to this one. 

>Also, what happens if we select an item with an empty value (similar to having an empty string as value for an option element in HTML)?

Since there isn't a value DOM attribute (just the prop for controlling the component), do you mean adding an empty string to `options.name`? Or something else? An empty string is valid for that and also for `value` in Ariakit.

Reply:
As a second consideration, I'm happy with the existing tests. Let's move forward and ship what we have, and we can always add another case if needed.